Improve transfered time from ort to torch

xadupre opened the pull request:
**Description**:

backward method creates many gradients stored in one OrtValueVector. The ownership of the data is then transfered to pytorch. This process needs the creation of a torch tensor for every OrtValue. One transfer costs:

* an allocation to a DLPack structure in ort
* the creation of a capsule to hold it
* the creation of a tensor in pytorch
* the allocation of another structure in pytorch
* the copy of a DLPack structure

The changes proposes to remove the first two steps. The DLPack structure is allocated on the stack and reused for every element of OrtValueVector, the capsule is also reused to save the creation of one python object. The cost of the transfer back to pytorch is approximatively reduced by 30%-50%.

All conversions from OrtValueVector to tuple of torch.tensor should go through function `_utils._ortvalues_to_torch_tensor`. If there is no boolean OrtValue, the function is 30% faster than the previous implementation.

**Motivation and Context**
- Performance: method forward and backward needs to take ownership of tensors from pytorch and give owernship of output or gradient. When there are many tensors to process (> 100 hundred), this can take a few milliseconds. This process cannot be efficiently parallelized. It creates many python object in a short time and goes through the GIL many time. Parallelization is inefficient in that case.

The conversion of 10 OrtValueVector of 100 OrtValue currently takes 0.003257s and 0.000871s with these changes. Measures obtained with the following script.

```
import time
import numpy 
from onnxruntime.capi._pybind_state import (
    OrtValue as C_OrtValue,
    OrtDevice as C_OrtDevice,
    OrtMemType, OrtValueVector)
from onnxruntime.capi import _pybind_state as C
import torch


def create_ortvalues(device=None):
    if device is None:
        device = C_OrtDevice(C_OrtDevice.cpu(), OrtMemType.DEFAULT, 0)
    memo = []
    ovect = OrtValueVector()
    for i in range(0, 100):
        vect = numpy.random.randn(2, 10).astype(numpy.float32)
        memo.append(vect)
        ort_value = C_OrtValue.ortvalue_from_numpy(vect, device)
        ovect.push_back(ort_value)
    return ovect, memo

def to_tensor_loop1(ovs):
    return tuple(torch.from_dlpack(ov.to_dlpack()) for ov in ovs)


def to_tensor_loop2(ovs):
    return tuple(torch._C._from_dlpack(ov.to_dlpack()) for ov in ovs)

def to_tensor_vector(ovs):
    return tuple(ovs.to_dlpacks(torch._C._from_dlpack))



fcts = [(to_tensor_loop1, None),
        (to_tensor_loop2, None),
        (to_tensor_vector, None)]


if False: # ort device
    # How to create a OrtValue with device == 'ort'?
    device_ort = C_OrtDevice(0, OrtMemType.DEFAULT, 0)
    print(device_ort.device_type())

    def ov2torch_eager(ovs):
        return tuple(C.to_aten_ort_device_tensor(ov) for ov in ovs)

    fcts.append((ov2torch_eager, device_ort))

for fct, dev in fcts:
    # check
    res = []
    for i in range(0, 10):
        ovs, memo = create_ortvalues(dev)
        begin = time.perf_counter()
        fct(ovs)
        end = time.perf_counter()
        delay = end - begin
        res.append(delay)

    print("%s: size=%d %f" % (fct.__name__, len(memo), sum(res)))
    print(res)
```

yuslepukhin commented on orttraining/orttraining/python/orttraining_pybind_state.cc:
```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
+
+        py::list res;
+
+        PyObject* capsule;
```
>PyObject* capsule [](http://example.com/codeflow?start=8&length=17)

Using py::capsule is very convenient and succinct.

yuslepukhin replied:
[https://stackoverflow.com/questions/54876346/pybind11-and-stdvector-how-to-free-data-using-capsules](https://stackoverflow.com/questions/54876346/pybind11-and-stdvector-how-to-free-data-using-capsules)

pengwa commented on orttraining/orttraining/python/orttraining_pybind_state.cc:
```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
+    #endif
+            if (obj == NULL)
+              throw std::runtime_error("Empty tensor returned.");
+            res.append(py::handle(obj));
```
do we need decrease obj ref cnt? assuming res.append() will increase the obj's ref_cnt by 1.

xadupre replied:
We need obj to persist after the execution of this function. obj is a torch tensor. It will be deleted when it is no longer needed later.

pengwa replied:
oh, I mean. after "obj = PyObject_CallOneArg" runs, obj as a PyObject* will have ref_cnt = 1. And the ref cnt will be decreased until we explicitly call Py_DECREF. 

On the other hand, I am not that sure, but have following guess: if we append obj into res, Py_REFCNT(obj) will be 2?  So after the function returns res, obj still have REF_CNT = 1. You can add add Py_REFCNT in multiple places to see the reference count.

xadupre replied:
You are right. PyList_Append increases the reference count.

xadupre replied:
However function PyObject_CallOneArg does not return a new reference, so I'm not sure about the count anymore. I used it to be faster. I'll remove it.

pengwa commented on orttraining/orttraining/python/orttraining_pybind_state.cc:
```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
+
+        if (tensor.Location().device.Type() == OrtDevice::CPU) {
+          OrtDLManagedTensor ort_dlmanaged_tensor;
+          capsule = PyCapsule_New(&ort_dlmanaged_tensor.tensor, "dltensor", NULL);
```
capsule  seems to be leaked after the whole function returns

xadupre replied:
true

iK1D commented on onnxruntime/python/onnxruntime_pybind_ortvalue.cc:
```diff
@@ -210,10 +210,18 @@ void addOrtValueMethods(pybind11::module& m) {
+      .def("__dlpack__", [](OrtValue* ort_value) -> py::object {
+        return py::reinterpret_steal<py::object>(ToDlpack(*ort_value));
+       }, "Returns a DLPack representing the tensor (part of __dlpack__ protocol).")
+      .def("__dlpack_device__", [](OrtValue* ort_value) -> py::tuple {
```
Maybe we can get this info from OrtValue directly instead of converting it to DLPack first.

xadupre replied:
We could do that if OrtValue was using DLPack structure to describe a tensor but it is not the case. So we need to create this intermediate structure. pytorch does the same: https://github.com/pytorch/pytorch/blob/master/torch/csrc/Module.cpp#L365, https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/DLConvertor.cpp#L217.

iK1D replied:
I think I am talking about the __dlpack_device__, PyTorch implements it here: https://github.com/pytorch/pytorch/blob/88d86de7d8887660e3277b06c49147c4e0df5b89/torch/_tensor.py#L1156. And it gets the info from the torch tensor (the self arugment) directly, though this is in Python code. Our code uses pybind and the code is in C++, but it's same thing.

xadupre replied:
I removed the creation of a new structure and only called the necessary code to return the device.

pengwa reviewed (commented):
talked with Vincent offline, here are few question about PyObject* ref_cnt management.

iK1D commented on onnxruntime/python/onnxruntime_pybind_ortvalue.cc:
```diff
@@ -210,10 +210,18 @@ void addOrtValueMethods(pybind11::module& m) {
         return FromDlpack(data.ptr(), is_bool_tensor);
-      })
+      }, "Convers a tensor from a external library into an OrtValue by means of the __dlpack__ protocol.")
+      .def("__dlpack__", [](OrtValue* ort_value) -> py::object {
```
PyTorch and some other platforms (such as Numpy)'s __dlpack__ also has a 'stream' argument, if it's CUDA/ROCM device, we need to sync the stream if streams are not same. Maybe we can skip such logic for now as we think that ORT and PT are running in same default stream, but maybe we still need to keep that argument here?

Note that PyTorch support such __dlpack__ and __dlpack_device__ from 1.10, currently we mainly run our ORTModule on 1.9.1, so this won't help if we use old PT version.

xadupre replied:
I looked into that, it is current written in python: https://github.com/pytorch/pytorch/blob/master/torch/_tensor.py#L1113. I need to move it to C++. We should keep the argument. It is not used now as this protocol is only used between ort and pytorch.

xadupre replied:
pytroch code about stream:

```
        if stream is not None and type(stream) is not int:
            # Stream pointers in CUDA/ROCm are uniquely numbered and can
            # be retrieved from their integer value.
            raise TypeError('stream must be ``int`` or ``none``')
        elif stream is not None and stream != -1:
            if self.device.type == 'cuda':
                stream = torch.cuda.streams.ExternalStream(stream)
                # Only synchronize on different streams
                if stream != torch.cuda.current_stream:
                    event = torch.cuda.Event()
                    event.record(torch.cuda.current_stream())
                    stream.wait_event(event)
```

iK1D commented on orttraining/orttraining/python/orttraining_pybind_state.cc:
```diff
@@ -345,6 +345,11 @@ std::unordered_map<std::string, std::unordered_map<std::string, py::object>> Con
   return py_tensor_state;
 }
 
+struct OrtDLManagedTensor {
```
Maybe we can move this definition from dlpack_converter.cc to dlpack_converter.h so we don't need to define this again here.

iK1D commented on orttraining/orttraining/python/orttraining_pybind_state.cc:
```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
         return py::reinterpret_steal<py::object>(ToDlpack(v->at(idx)));
-      });
+      })
+      .def("to_dlpack", [](const std::vector<OrtValue>& v, py::object to_tensor) -> py::list {
```
Actually I don't think the sharing one OrtDLManagedTensor works. The main purpose of OrtDLManagedTensor is to hold the "OrtValue handle"'s reference. It's passed to torch tensor by DLManagedTensor's manager_ctx and will be destroyed when torch tensor release by calling the DLManagedTensor's deleter. The destroy will decrease the ref count of that OrtValue. By saying passing tensor owership from ORT to PyTorch, we actually means passing the reference of the OrtValue from ORT to PyTorch (not the DLManagedTensor). Using only one OrtDLManagedTensor cannot achieve this as one OrtDLManagedTensor can hold only one Ortvalue. The idea here has two problem:

1) when using only one OrtDLManagedTensor, the OrtDLManagedTensor's handle (which is OrtValue type) is replaced by new value for each asignment, means that the ref count for previous OrtValue is decreased by 1. And the final one's ref count will also be decreased by 1 after this function when that single OrtDLManagedTensor is destroyed. If in the main thread we unref all these OrtValues (suppose it's the only place to hold them), then all the OrtValue will be destroyed, then the returned torch tensor will be invalid because they point to invalid data address. I.e., this idea actually fails to pass the ownership of OrtValue from ORT to PyTorch.

2) All the DLManagedTensor passed to different torch tensors have the same deleter, which will try to delete the same single OrtDLManagedTensor. But actually this shared OrtDLManagedTensor is destroyed when we go out of this function. So when we finish using those torch tensors and try to destroy them, the deleter will be called, all those deleter will try to delete the same pointer, an invalid pointer pointed to a destroyed OrtDLManagedTensor.

I don't think a single OrtDLManagedTensor can solve this problem. Another idea is an OrtDLManagedTensor pool. During the first step, we still need to create one OrtDLManagedTensor for each OrtValue output, but these OrtDLManagedTensor can be shared start from step 2 (if we think this allocation is a real problem).

xadupre replied:
You are right, I missed the fact the new tensor was responsible to delete the dlpack structure. So my local dlapck structure made onnxruntime crash because it could not be deleted. I modified the code to create a new dlpack structure for every tensor consuming it but I think it is still possible to create only one capsule instead of a capsule per dlpack structure.

iK1D commented on orttraining/orttraining/python/orttraining_pybind_state.cc:
```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
+            res.append(py::handle(obj));
+          }
+        } else {
+          // This piece of code should be removed when the issue on GPU is found.
```
What kind of issue here?

xadupre replied:
When an OrtValue is created, it copied whatever needs to be copied but the deleter attached the ortvalue deletes the data pointer and the DLPack structure used to create it even if it is not needed anymore. I first assumed the python capsule was responsible for that but it only is when the capsule is not consumed. Pytorch does the same thing.

```
OrtValue DlpackToOrtValue(DLManagedTensor* dlpack, bool is_bool_tensor) {
  // ORT only supports contiguous tensor for now.
  ORT_ENFORCE(IsContiguousTensor(dlpack->dl_tensor), "ORT only supports contiguous tensor for now.");
  OrtDevice device = GetOrtDevice(dlpack->dl_tensor.device);
  MLDataType data_type = GetOrtValueDataType(dlpack->dl_tensor.dtype, is_bool_tensor);
  OrtMemoryInfo info(GetOrtDeviceName(device), OrtDeviceAllocator, device, device.Id());
  std::unique_ptr<Tensor> p_tensor = std::make_unique<Tensor>(
      data_type, TensorShape(dlpack->dl_tensor.shape, static_cast<size_t>(dlpack->dl_tensor.ndim)),
      dlpack->dl_tensor.data, info);

  OrtValue ort_value;
  std::function<void(void*)> deleter = [dlpack](void* p) {
    if (dlpack->deleter != NULL)
      dlpack->deleter(dlpack);
    auto deleter = DataTypeImpl::GetType<Tensor>()->GetDeleteFunc();
    if (deleter != NULL)
      deleter(p);
  };

  ort_value.Init(p_tensor.release(), DataTypeImpl::GetType<Tensor>(), deleter);
  return ort_value;
}
```


```
void DlpackCapsuleDestructor(PyObject* data) {
  DLManagedTensor* dlmanaged_tensor = reinterpret_cast<DLManagedTensor*>(PyCapsule_GetPointer(data, "dltensor"));
  if (dlmanaged_tensor) {
    // The dlmanaged_tensor has not been consumed, call deleter ourselves.
    dlmanaged_tensor->deleter(const_cast<DLManagedTensor*>(dlmanaged_tensor));
  } else {
    // The dlmanaged_tensor has been consumed,
    // PyCapsule_GetPointer has set an error indicator.
    PyErr_Clear();
  }
}
```

iK1D commented on orttraining/orttraining/python/orttraining_pybind_state.cc:
```diff
@@ -357,16 +358,95 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
         return py::reinterpret_steal<py::object>(ToDlpack(v->at(idx)));
-      });
+      })
+      .def("to_dlpack", [](const std::vector<OrtValue>& v, py::object to_tensor) -> py::list {
```
Rename "to_dlpack" to "to_tensors_of_target_framework" or similar? Because the type of result here depends on the "to_tensor" function passed in.

xadupre replied:
I changed the behaviour of the function, if to_tensor is None, then it returns a capsule for all OrtValue in the vector. I think dlpack should be used somewhere in the name as it relies on it.

iK1D commented on onnxruntime/core/dlpack/dlpack_converter.h:
```diff
@@ -10,13 +10,28 @@
+// similar to the previous function but the structure DLManagedTensor
+// is already allocated. This function can be used when multiple
+// OrtValue are converted in the same function. This saves allocations.
+DLManagedTensor* OrtValueToDlpack(OrtValue& ort_value, OrtDLManagedTensor* ort_dlmanaged_tensor, DlpackDeleterFct* deleter);
```
This one is not used in your latest change? And all the changes in dlpack_converter.h/cc are all related to this new added function?

xadupre replied:
Removed.

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -50,6 +51,33 @@ def _ortvalue_to_torch_tensor(ortvalue, device):
     dlpack_tensor = ortvalue.to_dlpack()
```
Since you replace all _ortvalue_to_torch_tensor by _ortvalues_to_torch_tensor, maybe we can remove the old one.

xadupre replied:
I let it just in case somebody needs to convert only one ortvalue. I can remove it.

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -50,6 +51,33 @@ def _ortvalue_to_torch_tensor(ortvalue, device):
+        return tuple()
+    if (hasattr(ortvalues, 'to_dlpack') and
+            all(ortvalue.data_type() != 'tensor(bool)' for ortvalue in ortvalues)):
+        if device == 'ort':
```
check device is str or use device.type?

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -50,6 +51,33 @@ def _ortvalue_to_torch_tensor(ortvalue, device):
+    if len(ortvalues) == 0:
+        return tuple()
+    if (hasattr(ortvalues, 'to_dlpack') and
+            all(ortvalue.data_type() != 'tensor(bool)' for ortvalue in ortvalues)):
```
Why do we exclude bool here? I think we can still use the new way to generate the torch tensor tuples and then apply the conversion by checking tensor element type after that.

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -50,6 +51,33 @@ def _ortvalue_to_torch_tensor(ortvalue, device):
+            all(ortvalue.data_type() != 'tensor(bool)' for ortvalue in ortvalues)):
+        if device == 'ort':
+            return tuple(ortvalues.to_dlpack(
+                lambda dlpack_structure: C_OrtValue.from_dlpack(dlpack_structure, False)))
```
Should be C.ort_from_dlpack here according to the _torch_tensor_from_dl_pack. C.ort_from_dlpack returns torch tensor, while C_OrtValue.from_dlpack returns OrtValue.

iK1D reviewed (commented):
I tested the changes using ULR-XL. Before the changes it takes ~4ms after backward, while after the changes, it's ~2.5ms. The changes indeed help to improve the perf. Besides this, we can remove some common code to outside of training steps (in one of my PR), that can also help to reduce the time a little.

I think it's better to run memory leak test to make sure there is no leak before we can push to master.

iK1D commented on onnxruntime/core/dlpack/dlpack_converter.cc:
```diff
@@ -75,37 +75,6 @@ DLDataType GetDlpackDataType(const OrtValue& ort_value) {
-  return device;
-}
-
-struct OrtDLManagedTensor {
```
The new changed code doesn't use OrtDLManagedTensor or DlpackDeleter, maybe we don't need to move these two pieces. OrtDLManagedTensor doesn't need to add to header file.

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -38,17 +39,74 @@ def _ortvalue_from_torch_tensor(torch_tensor):
+            all(ortvalue.proto_type() != 9 for ortvalue in ortvalues)):
+        if 'ort' == (device if isinstance(device, str) else device.type):
+            return tuple(ortvalues.to_dlpack(
+                lambda dlpack_structure: C.OrtValue.from_dlpack(dlpack_structure, False)))
```
I still don't quite understand why "C.OrtValue.from_dlpack" here. This function should returns torch tensor list, but C.OrtValue.from_dlpack return OrtValue. And the original code uses "C.ort_from_dlpack", which is different. There are one more place below using "C.OrtValue.from_dlpack", which will also return OrtValue list instead of torch tensor list.

xadupre commented:
https://github.com/microsoft/onnxruntime/pull/9610#discussion_r749948910: the previous implementation was supporting that case. An OrtValue could be consumed by another OrtValue. I can remove it and see if one unit test fails.

iK1D commented:
> [#9610 (comment)](https://github.com/microsoft/onnxruntime/pull/9610#discussion_r749948910): the previous implementation was supporting that case. An OrtValue could be consumed by another OrtValue. I can remove it and see if one unit test fails.

Does previous implementation means below code? What kind of arguments will trigger the function returing OrtValue instead of torch tensor? The problem is I don't think we have UT covering the device.type == 'ort' case.

def _torch_tensor_from_dl_pack(dlpack, ortvalue, device):
    torch_tensor = from_dlpack(dlpack) if device.type != 'ort' else C.ort_from_dlpack(dlpack)
    return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor

xadupre commented:
> > [#9610 (comment)](https://github.com/microsoft/onnxruntime/pull/9610#discussion_r749948910): the previous implementation was supporting that case. An OrtValue could be consumed by another OrtValue. I can remove it and see if one unit test fails.
> 
> Does previous implementation means below code? What kind of arguments will trigger the function returing OrtValue instead of torch tensor? The problem is I don't think we have UT covering the device.type == 'ort' case.
> 
> def _torch_tensor_from_dl_pack(dlpack, ortvalue, device): torch_tensor = from_dlpack(dlpack) if device.type != 'ort' else C.ort_from_dlpack(dlpack) return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor

I assume this code was added because it was needed. I don't know if there was a unit test covering this case but I added one here: https://github.com/microsoft/onnxruntime/pull/9610/files#diff-ce379df71758bffe5213b2de9b721c7cdf9b8ea4df1e9baa6a00826e0984f86eR179.

iK1D commented:
> > > [#9610 (comment)](https://github.com/microsoft/onnxruntime/pull/9610#discussion_r749948910): the previous implementation was supporting that case. An OrtValue could be consumed by another OrtValue. I can remove it and see if one unit test fails.
> > 
> > 
> > Does previous implementation means below code? What kind of arguments will trigger the function returing OrtValue instead of torch tensor? The problem is I don't think we have UT covering the device.type == 'ort' case.
> > def _torch_tensor_from_dl_pack(dlpack, ortvalue, device): torch_tensor = from_dlpack(dlpack) if device.type != 'ort' else C.ort_from_dlpack(dlpack) return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor
> 
> I assume this code was added because it was needed. I don't know if there was a unit test covering this case but I added one here: https://github.com/microsoft/onnxruntime/pull/9610/files#diff-ce379df71758bffe5213b2de9b721c7cdf9b8ea4df1e9baa6a00826e0984f86eR179.

Can you add one more assert in def _ortvalues_to_torch_tensor_ortvaluevector(self, device) at the end? Something like:

```
for t in tensors:
    assert(isinstance(t, torch.Tensor))
```

I think this assert should be passed when using the old code. But will fail on the new code.

`self.assertEqual(ptr, [t.data_ptr() for t in tensors])` works because both torch.Tensor and OrtValue has data_ptr() and returns the data pointer.

SherlockNoMad commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -38,17 +39,74 @@ def _ortvalue_from_torch_tensor(torch_tensor):
-def _torch_tensor_from_dl_pack(dlpack, ortvalue, device):
-    torch_tensor = from_dlpack(dlpack) if device.type != 'ort' else C.ort_from_dlpack(dlpack)
-    return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor
+def _ortvalues_to_torch_tensor(ortvalues, device):
```
This function need to return a pytorch tensor. 
In the original code,  "C.ort_from_dlpack(dlpack)" actually returns a pytorch tensor. (I know the naming here is confusing, but it's returning a at::Tensor).
In the modified code, "C.OrtValue.from_dlpack()" is actually returning a ORTValue.

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -38,17 +39,74 @@ def _ortvalue_from_torch_tensor(torch_tensor):
+            tensors.append(te)
+        return tuple(tensors)
+
+    if 'ort' == (device if isinstance(device, str) else device.type):
```
I would like to clarify more on this. You are changing `_ortvalue_to_torch_tensor` to `_ortvalues_to_torch_tensor`. From the function name it means in any case we should return torch.Tensor (list) at the end. But in your new code, it returns an OrtValue list at line 48-49, and here, it return torch.Tensor list if `hasattr(C, 'ort_from_dlpack')` is true, but will still return OrtValue list if false at line 98-99. The main problem is `C.ort_from_dlpack` and `C.OrtValue.from_dlpack` are two different functions, one returns torch.Tensor while the other returns OrtValue. Just as Sherlock said, the names are comfusing, but they are different.

'ort' device type is a special device type, it's used to descibe the device of original OrtValue, just like 'cpu' and 'cuda', but not the target tensor type we need to generate. For this function, we need to always return torch.Tensor. 'ort' device is now only used in ort_eager mode. As I said previously, we don't have any UT or CI pipeline to cover this for now. if you want to try it, you need to use specific PyTorch version and build ORT using "--build_eager_mode" to enable it. You can run the test script at 'orttraining/orttraining/eager/test_model/mnist_fc_training.py' if you can have the ort_eager mode environment setup, but this is very hard to do for now. Acutally you should not use `if hasattr(C, 'ort_from_dlpack')` here to do different things. 'ort' device is for ort_eager mode, and ort_eager mode needs "--build_eager_mode", and when this is enabled, the `C` must have `ort_from_dlpack`. So the correct behavior here should be to `assert(hasattr(C, 'ort_from_dlpack'))` if device type is 'ort'. Your changes acutally break the ort_eager mode, which is used and developed by someone else.

In your new test code, I can see you added `isinstance(v, tensor_type)` but still pass `C_OrtValue` as the target type if device is 'ort': `self._ortvalues_to_torch_tensor_ortvaluevector('ort', C_OrtValue)`, which is logically wrong. As mentioned above, in any case, the returned tensor should be torch.Tensor. I think for any UT, no matter old one or newly added, we should run it on both the code before and after the changes. The new tests will definitely work on your new changed code, but you should also try to roll back your changes to old version and run the tests to see if they can pass, if not, the tests are not valid. But I guess you will not be able to do this for 'ort' device, as I mentioned above, it requires specific environment, which is not easy to setup, and normally we don't enable "--build_eager_mode" when building ORT. Any code using device 'ort' without "--build_eager_mode" is not valid.

iK1D commented on onnxruntime/python/onnxruntime_pybind_ortvalue.cc:
```diff
@@ -213,12 +228,37 @@ void addOrtValueMethods(pybind11::module& m) {
       ;
+
+#ifdef ENABLE_TRAINING
+  m.def("may_dlpack_bool_tensor", [](py::capsule cap) -> bool {
```
Why do we need this function? And it's confusing from the name and the doc here. The condition here means it's possible to be UINT8 or BOOL, so the "may" in the name seems reasonable, but the doc "Tells if a DLPack structure is a boolean tensor." is confusing. This function is used only once (ignoring the test file) at "_utils.py" below: `if C.may_dlpack_bool_tensor(dlp) and ortvalues[i].proto_type() == 9:`, but it's not possible that proto_type is 9 but may_dlpack_bool_tensor returns false. We should not use IF to hide problem, if we really want to do some check, we should assert.

iK1D commented on orttraining/orttraining/python/orttraining_pybind_state.cc:
```diff
@@ -359,14 +360,76 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
         return py::reinterpret_steal<py::object>(ToDlpack(v->at(idx)));
-      });
+      })
+      .def("to_dlpack", [](const std::vector<OrtValue>& v, py::object to_tensor) -> py::list {
```
I still think we should rename this one. Even we still want to keep the dlpack, it should also be "to_dlpacks".

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -38,18 +39,46 @@ def _ortvalue_from_torch_tensor(torch_tensor):
+        return tuple()
+    if hasattr(ortvalues, 'to_dlpack'):
+        if all(ortvalue.proto_type() != 9 for ortvalue in ortvalues):
+            if 'ort' == device.type and hasattr(C, 'ort_from_dlpack'):
```
I suggest not to use IF clause to return different torch tensors when C has ort_from_dlpack or not. I understand that you make it this way because you still want to add some tests related to 'ort' device, but `C.ort_from_dlpack` and torch's _from_dlpack return different torch.Tensors. `C.ort_from_dlpack` returns a torch.Tensor that use `ORTTensorImpl` as the final implementation, which is written by ourselves, but torch's _from_dlpack returns a normal torch.Tensor implemented by torch. If the device is 'ort', returning a normal torch.Tensor (using torch's _from_dlpack) will cause failure when it goes to somewhere else in the backend at the end. So actually this IF is hiding some problem (why C doesn't have ort_from_dlpack when device is 'ort') here and may bring trouble to debugging. Assertion (that C must have ort_from_dlpack) should be the correct behavior here. (The original code calls `C.ort_from_dlpack` directly, and if C doesn't have ort_from_dlpack, Python will report the problem. It's better we don't introduce any new logic compared to the old code.) @souptc who is the developer of ort_eager mode may want to add more comment. Same comment applies to below code from line 54 to 57.

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -38,18 +39,46 @@ def _ortvalue_from_torch_tensor(torch_tensor):
+    if len(ortvalues) == 0:
+        return tuple()
+    if hasattr(ortvalues, 'to_dlpack'):
+        if all(ortvalue.proto_type() != 9 for ortvalue in ortvalues):
```
If we agree that the new 'to_dlpack' works well and fast, then even ortvalue.proto_type() is 9, we can still use the new 'to_dlpack' to get all torch.Tensors, then convert those torch tensors at position that their corresponding OrtValues are bool to bool torch tensors. So we don't need the `if hasattr(ortvalues, 'dlpack_at'):` clause below anymore. Also suggest use Enum name instead of number (9) for better code reading.

SherlockNoMad commented on onnxruntime/core/dlpack/dlpack_converter.cc:
```diff
@@ -230,8 +226,11 @@ OrtValue DlpackToOrtValue(DLManagedTensor* dlpack, bool is_bool_tensor) {
   std::function<void(void*)> deleter = [dlpack](void* p) {
-    dlpack->deleter(dlpack);
-    DataTypeImpl::GetType<Tensor>()->GetDeleteFunc()(p);
+    if (dlpack->deleter != NULL)
```
Agree, we should crash the process if either deleter is nullprt. 
A non-existing delete implies memory leak.

yuslepukhin commented on onnxruntime/python/onnxruntime_pybind_mlvalue.cc:
```diff
@@ -113,8 +113,10 @@ OrtMemoryInfo GetMemoryInfoPerDeviceType(const OrtDevice& ort_device) {
 int32_t GetProtoType(const OrtValue& ort_value) {
   if (ort_value.IsTensor()) {
     return ort_value.Get<Tensor>().GetElementType();
+#if !defined(ORT_MINIMAL_BUILD)
```
This may also benefit of `#if !defined(DISABLE_SPARSE_TENSORS)`

xadupre commented:
Hi @SherlockNoMad, should I bring other changes on this review?

iK1D commented on onnxruntime/core/dlpack/dlpack_converter.cc:
```diff
@@ -249,8 +250,11 @@ OrtValue DlpackToOrtValue(DLManagedTensor* dlpack, bool is_bool_tensor) {
     dlpack->deleter(dlpack);
-    DataTypeImpl::GetType<Tensor>()->GetDeleteFunc()(p);
+    auto deleter = DataTypeImpl::GetType<Tensor>()->GetDeleteFunc();
+    if (deleter != NULL)
```
deleter can't be NULL, can we ORT_ENFORCE here?

iK1D commented on onnxruntime/python/onnxruntime_inference_collection.py:
```diff
@@ -604,6 +604,13 @@ def data_type(self):
         '''
         return self._ortvalue.data_type()
 
+    def proto_type(self):
```
Can we change the name to element_type or data_type? I think the proto_type is confusion. Same for other places.

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -38,21 +40,65 @@ def _ortvalue_from_torch_tensor(torch_tensor):
+    if len(ortvalues) == 0:
+        return tuple()
+    if hasattr(ortvalues, 'to_dlpacks'):
+        if all(ortvalue.proto_type() != TensorProto.BOOL for ortvalue in ortvalues):
```
Is it possible to move all the BOOL handling code to the end of the function? For example, not return directly for each IF clause, but get a tensor_list without handling BOOL tensor. And at the end of this function, add:

return tuple([tensor.to(torch.bool) if ortvalues[i].proto_type() == TensorProto.BOOL else tensor for i, tensor in enumerate(tensor_list)])

With this way, we can also use the new to_dlpacks functions for a ortvalue list that also contains bool tensor, because it has better perf?

xadupre commented:
I introduced two changes:
* a method has_bool_tensor to check a OrtValueTensor has one boolean tensor.
* the conversion of OrtValueTensor into torch tensor uses `to_dlpacks` every time (except for `device == 'ort'`), then the method postprocesses the boolean tensors with `.to(torch.bool)`.

The method is 2x faster the previous implementation in this PR. Yellow is the implementation before this change (https://github.com/microsoft/onnxruntime/pull/9610/commits/f90ff2431c7280400f7ce06fcbd671f4659ce6d1), orange after.

![image](https://user-images.githubusercontent.com/22452781/161073769-0abcb792-31b0-4375-b348-2ece18fb53d6.png)

```python
import time
import pprint
import numpy 
from onnx import TensorProto
import pandas
from tqdm import tqdm
from onnxruntime.capi._pybind_state import (
    OrtValue as C_OrtValue,
    OrtDevice as C_OrtDevice,
    OrtMemType, OrtValueVector)
from onnxruntime import OrtDevice
from onnxruntime.capi import _pybind_state as C
from onnxruntime.training.ortmodule._utils import _ortvalues_to_torch_tensor
import torch
from torch._C import _from_dlpack
from torch.utils.dlpack import from_dlpack


def create_ortvalues(dtype, N, device=None):
    if device is None:
        device = C_OrtDevice(C_OrtDevice.cpu(), OrtMemType.DEFAULT, 0)
    memo = []
    ovect = OrtValueVector()
    for i in range(0, N):
        vect = numpy.random.randn(2, 10).astype(dtype)
        memo.append(vect)
        ort_value = C_OrtValue.ortvalue_from_numpy(vect, device)
        ovect.push_back(ort_value)
    return ovect, memo


def to_tensor_loop1(ovs):
    return tuple(torch.from_dlpack(ov.to_dlpack()) for ov in ovs)


def to_tensor_loop2(ovs):
    return tuple(torch._C._from_dlpack(ov.to_dlpack()) for ov in ovs)


def to_tensor_vector(ovs):
    return tuple(ovs.to_dlpacks(_from_dlpack))


cpu = OrtDevice(C_OrtDevice(C_OrtDevice.cpu(), OrtMemType.DEFAULT, 0))
cpu = torch.device('cpu')


def this_ortvalues_to_torch_tensor(ortvalues):
    device = cpu
    if len(ortvalues) == 0:
        return tuple()
    if hasattr(ortvalues, 'to_dlpacks'):
        if all(ortvalue.element_type() != TensorProto.BOOL for ortvalue in ortvalues):
            if 'ort' == device.type:
                if not hasattr(C, 'to_aten_ort_device_tensor'):
                    raise AttributeError("onnxruntime is missing to_aten_ort_device_tensor needed to support device == 'ort'.")
                # The below expression could be used as well but it fails.
                # return tuple(ortvalues.to_dlpacks(C.to_aten_ort_device_tensor))
                # The following expressions fails too:
                # tensors = tuple(C.to_aten_ort_device_tensor(ov) for ov in ortvalues);print(tensors[0])
                # due to: NotImplementedError: Could not run 'aten::_cat' with arguments from the 'ORT' backend.
                # C.to_aten_ort_device_tensor returns a torch tensor with device 'ort:0'
                # instead of one of the allowed backend (cpu, cuda:0, ...).
                return tuple(C.to_aten_ort_device_tensor(ov) for ov in ortvalues)
            return tuple(ortvalues.to_dlpacks(_from_dlpack))

        # DLPack structure does not know for sure if it stores boolean
        # or uint8. Method to_dlpacks cannot be used in that case.
        # Signature of *dl_packs* is `to_dlpacks(dlp, fct) -> list[torch.Tensor]`.
        # And fct is a function with signature `fct(dlp) -> torch.Tensor`.
        # Boolean tensors are converted into uint8 tensor with the DLPack protocol.
        # Therefore, the function `fct` does not know if the dlpack structure
        # is a boolean tensor or a uint8 tensor.
        # We could either consider another function as an input in
        # `to_dlpacks` or add an argument to `fct(dlp, ortvalue)`.
        # Second option makes it impossible to directly use `_from_dlpack` or
        # or `from_dlpack` from torch.
        # The best option would be to add boolean type in DLDataTypeCode.

    # if ortvalues[0].data_type() != 'tensor(bool)':
    #     raise AssertionError(ortvalues[0].data_type())
    if 'ort' == device.type:
        if not hasattr(C, 'to_aten_ort_device_tensor'):
            raise AttributeError("onnxruntime is missing to_aten_ort_device_tensor needed to support device == 'ort'.")
        return tuple(C.to_aten_ort_device_tensor(ov) for ov in ortvalues)

    if hasattr(ortvalues, 'dlpack_at'):
        tensors = []
        for i in range(len(ortvalues)):
            dlp = ortvalues.dlpack_at(i)
            te = from_dlpack(dlp)
            if C.is_dlpack_uint8_tensor(dlp) and ortvalues[i].element_type() == TensorProto.BOOL:
                te = te.to(torch.bool)
            tensors.append(te)
        return tuple(tensors)

    tensors = []
    for i in range(len(ortvalues)):
        ov = ortvalues[i]
        if hasattr(ov, '_ortvalue'):
            ov = ov._ortvalue
        dlp = ov.to_dlpack()
        te = from_dlpack(dlp)
        if ov.element_type() == TensorProto.BOOL:
            te = te.to(torch.bool)
        tensors.append(te)
    return tuple(tensors)


def ort_to_torch_tensor2(ortvalues):
    device = cpu
    if len(ortvalues) == 0:
        return tuple()
    if 'ort' == device.type:
        if not hasattr(C, 'to_aten_ort_device_tensor'):
            raise AttributeError("onnxruntime is missing to_aten_ort_device_tensor needed to support device == 'ort'.")
        return tuple(C.to_aten_ort_device_tensor(ov) for ov in ortvalues)
    res = ortvalues.to_dlpacks(_from_dlpack)
    for i in range(0, len(ortvalues)):
        if ortvalues.element_type_at(i) == TensorProto.BOOL:
            res[i] = res[i].to(torch.bool)
    return tuple(res)


def ort_to_torch_tensor3(ortvalues):
    device = cpu
    if len(ortvalues) == 0:
        return tuple()
    if 'ort' == device.type:
        if not hasattr(C, 'to_aten_ort_device_tensor'):
            raise AttributeError("onnxruntime is missing to_aten_ort_device_tensor needed to support device == 'ort'.")
        return tuple(C.to_aten_ort_device_tensor(ov) for ov in ortvalues)
    res = ortvalues.to_dlpacks(_from_dlpack)
    if ortvalues.has_bool_tensor():
        for i in range(0, len(ortvalues)):
            if ortvalues.element_type_at(i) == TensorProto.BOOL:
                res[i] = res[i].to(torch.bool)
    return tuple(res)


def ort_ortvalues_to_torch_tensor(t):
    return _ortvalues_to_torch_tensor(t, cpu)

fcts = []

for i in (1, 2, 3, 4, 5, 10, 20, 50, 100, 500, 1000):
    fcts.extend([
        dict(fct=to_tensor_loop1, device=None, n_tensors=i, dtype=numpy.float32),
        dict(fct=to_tensor_loop2, device=None, n_tensors=i, dtype=numpy.float32),
        dict(fct=to_tensor_vector, device=None, n_tensors=i, dtype=numpy.float32),
        dict(fct=ort_ortvalues_to_torch_tensor, device=None, n_tensors=i, dtype=numpy.float32),
        dict(fct=this_ortvalues_to_torch_tensor, device=None, n_tensors=i, dtype=numpy.float32),
        dict(fct=ort_to_torch_tensor2, device=None, n_tensors=i, dtype=numpy.float32),
        dict(fct=ort_to_torch_tensor3, device=None, n_tensors=i, dtype=numpy.float32),
    ])
    fcts.extend([
        dict(fct=to_tensor_loop1, device=None, n_tensors=i, dtype=numpy.bool_),
        dict(fct=to_tensor_loop2, device=None, n_tensors=i, dtype=numpy.bool_),
        dict(fct=to_tensor_vector, device=None, n_tensors=i, dtype=numpy.bool_),
        dict(fct=ort_ortvalues_to_torch_tensor, device=None, n_tensors=i, dtype=numpy.bool_),
        dict(fct=this_ortvalues_to_torch_tensor, device=None, n_tensors=i, dtype=numpy.bool_),
        dict(fct=ort_to_torch_tensor2, device=None, n_tensors=i, dtype=numpy.bool_),
        dict(fct=ort_to_torch_tensor3, device=None, n_tensors=i, dtype=numpy.bool_),
    ])


if False: # ort device
    # How to create a OrtValue with device == 'ort'?
    device_ort = C_OrtDevice(0, OrtMemType.DEFAULT, 0)
    print(device_ort.device_type())

    def ov2torch_eager(ovs):
        return tuple(C.to_aten_ort_device_tensor(ov) for ov in ovs)

    fcts.append((ov2torch_eager, device_ort))

times = {}
data = []
for index, obs in enumerate(tqdm(fcts)):
    fct = obs['fct']
    # check
    res = []
    for i in range(0, 100):
        ovs, memo = create_ortvalues(obs['dtype'], obs['n_tensors'], obs['device'])
        begin = time.perf_counter()
        fct(ovs)
        end = time.perf_counter()
        delay = end - begin
        res.append(delay)

    obs = obs.copy()
    obs['index'] = index
    obs['name'] = fct.__name__
    obs['dim'] = len(memo)
    obs['n_run'] = len(res)
    obs['time'] = sum(res)
    obs['dtype'] = str(obs['dtype']).split('.')[-1].strip("'_>'")
    data.append(obs)
    times[obs['name'], obs['dim'], obs['dtype']] = obs['time']
    
for obs in data:
    obs['ratio'] = obs['time'] / times['to_tensor_vector', obs['dim'], obs['dtype']]

df = pandas.DataFrame(data).set_index('index')
print(df.head(n=2).T)
piv = pandas.pivot_table(df, index=['dtype', 'name'], columns=['n_tensors'], values=['time', 'ratio'])
print(piv)
df.to_excel('data.xlsx')
piv.to_excel('piv.xlsx')
pprint.pprint(data)
```

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -65,8 +65,15 @@ def _ortvalues_to_torch_tensor(ortvalues, device):
+    else:
+        # This happens because TrainingAgent.run_forward, TrainingAgent.run_backward
+        # do not use OrtValueVector.
+        res = [from_dlpack(ov.to_dlpack()) for ov in ortvalues]
```
Could we change it to use `OrtValueVector`? This else branch is slower than the if one, so maybe the API change is justified by the performance gain

xadupre commented:
I thought of doing that. We should replace std::vector<OrtValue> by OrtValueVector in run_forward and run_backward but I was wondering if these two functions were used by other projects or teams. In that case, I think doing this change in another is safer.

thiagocrepaldi commented:
> I thought of doing that. We should replace std::vector by OrtValueVector in run_forward and run_backward but I was wondering if these two functions were used by other projects or teams. In that case, I think doing this change in another is safer.

they were implemented for ORTModule , specifically. It is safe to change it

iK1D commented on orttraining/orttraining/python/training/ortmodule/_utils.py:
```diff
@@ -56,21 +58,43 @@ def _ortvalue_from_torch_tensor(torch_tensor):
+        return tuple(C.to_aten_ort_device_tensor(ov) for ov in ortvalues)
+    if hasattr(ortvalues, 'to_dlpacks'):
+        res = ortvalues.to_dlpacks(_from_dlpack)
+        has_bool_tensor = ortvalues.has_bool_tensor()
```
Maybe we can change has_bool_tensor() to bool_tensor_indices() to return all the indices of bool tensors in the list? So below IF is len(indices) > 0, and we can use this indices directly instead of another FOR loop to check the flag again.

iK1D commented:
> > I thought of doing that. We should replace std::vector by OrtValueVector in run_forward and run_backward but I was wondering if these two functions were used by other projects or teams. In that case, I think doing this change in another is safer.
> 
> they were implemented for ORTModule , specifically. It is safe to change it

If it's save to use OrtValueVector, then we don't need the ELSE clouse at all, because OrtValueVector must have "to_dlpacks" defined.

xadupre commented:
> > > I thought of doing that. We should replace std::vector by OrtValueVector in run_forward and run_backward but I was wondering if these two functions were used by other projects or teams. In that case, I think doing this change in another is safer.
> > 
> > 
> > they were implemented for ORTModule , specifically. It is safe to change it
> 
> If it's save to use OrtValueVector, then we don't need the ELSE clouse at all, because OrtValueVector must have "to_dlpacks" defined.

I hope. Otherwise, that means I'll have more code to refactor.

xadupre commented:
It turns out it is difficult to introduce one OrtValueVector instead of a list before converting it into torch tensor. The reason is this list comes from an API defined in onnruntime (`IOBinding.GetOutputs`). It is a `std::vector<OrtValue>` translated by pybind11 into a list. OrtValueVector is defined in onnxruntime-training and is also a `std::vector<OrtValue>` but translated into OrtValueVector by pybind11. From a C++ point of view, it is the same object. From python, one is a list, the other one is closer to a capsule. So I introduced `_ortvalues_to_torch_tensor_list` working only with list, and `_ortvalues_to_torch_tensor` only works with OrtValueVector. That makes it easy for a future refactoring. My recommendation would be to move OrtValueVector to onnxruntime but that's a bigger change.

thiagocrepaldi reviewed (approved):
Unblocking merge, but we discussed offline for another solution refactoring IObinding
